include/mbedtls/lms.h

```diff
@@ -0,0 +1,254 @@
+/**
+ * \file lms.h
+ *
+ * \brief This file provides an API for the LMS post-quantum-safe stateful-hash
+ *        public-key signature scheme.
+ */
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#ifndef MBEDTLS_LMS_H
+#define MBEDTLS_LMS_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "mbedtls/private_access.h"
+#include "mbedtls/lmots.h"
+
+#define MBEDTLS_ERR_LMS_BAD_INPUT_DATA   -0x0011 /**< Bad data has been input to an LMS function */
+#define MBEDTLS_ERR_LMS_OUT_OF_PRIV_KEYS -0x0013 /**< Specified LMS key has utilised all of its private keys */
+#define MBEDTLS_ERR_LMS_VERIFY_FAILED    -0x0015 /**< LMS signature verification failed */
+#define MBEDTLS_ERR_LMS_ALLOC_FAILED     -0x0017 /**< LMS failed to allocate space for a private key */
+
+#define MBEDTLS_LMS_TYPE_LEN        (4)
+#define MBEDTLS_LMS_H_TREE_HEIGHT       (10)
+#define MBEDTLS_LMS_M_NODE_BYTES        (32)
+
+#define MBEDTLS_LMS_SIG_LEN (MBEDTLS_LMOTS_Q_LEAF_ID_LEN + MBEDTLS_LMOTS_SIG_LEN + \
```

Comment:
On the one hand, these values are hard-coded constants, yet we pass the algorithm in as an enum (which has one definition). How will these constants, and indeed the API for export/import of the pubkey work if we support different variants?

Reply:
The reason is that some of the macros are used to allocate buffers. We could make the algorithm a macro as well, but since it's a function argument I figured giving a type hint to callers was worth it.

WRT how this works with multiple variants, the biggest issue is memory allocation. We'd have to switch a bunch of things over to dynamic allocation (which is fine I guess, mbedtls allows that). I put it in this way to start with just for simplicity's sake. I guess what'd happen is you'd have some macro/function which returned the H/M/etc values when passed a parameter set, that seems easiest.

Import / export should be unaffected, except for slight changes in size.

Reply:
It wasn't really that the algorithm was an enum, that's fine, it is that we have an algorithm that looks like it could be given different values, yet the implementation will only work with a single value. It isn't that significant, given that we only support a single algorithm, but supporting an additional one would require more work than might be obvious from seeing the argument passed to the function.

Reply:
Hmm yes, you're right this is currently potentially confusing. As I see it we have 3 options:
1) Change the LMS/LMOTS functions to not support a parameter type
2) Keep the current system
3) Implement a function that maps a type parameter to the required constant values. Doing this would likely require dynamic allocation, so probably a similar amount of work to _actually_ supporting most of the parameter sets (lmots parameters with N<8 are tricky as you end up needing to either allocate more memory or operate on sub-byte sizes).

I was avoiding 3, since adding more complexity is only going to make this review process harder, though I think it's a likely upgrade in future if anyone ever ends up needing it. 1 seems like a strange choice, since we're removing functionality, and we'd need to put it back if we ever ended up supporting multiple parameter sets.

Hence the selection of the current approach. Do you think one of the others would be better (or is there one I've missed)?